### ISSUE TITLE
"certbot certificates --json"

### ISSUE THREAD

**Issue by pde** [certbot/certbot#3909]
The `certificates` subcommand should have a `--json` flag for output that's easily ingestible by other programs.

**Comment by dashaxiong** [comment]
@pde Would the following output format be a good place to start?
A list of certificate dicts. Each dict has a certificate name, a list of domains, an expiry date, certificate path, and private key path. As per current output of the certificates subcommand. 

Formatted as in this gist:
[proposed json output](https://gist.github.com/dashaxiong/6731ec8debfc3d51e80ca9dcb4694963)

**Comment by dashaxiong** [comment]
PR #4138 adds this feature. Perhaps hold off on merging it until the Grep feature (#3910) has been added also, then merge just the PR (yet to be completed) for the Greppable output, which will include the JSON output feature.

Be great if the PR 4138 can be reviewed.

**Comment by bmw** [comment]
PR in #4138.

**Comment by bmw** [comment]
PR hasn't seen any updates in a while, but the code from it is available at https://github.com/certbot/certbot/tree/dashaxiong-json-certificate-output.

**Comment by grenade** [comment]
if you end up here because you just want to get json output from `certbot certificates`, you can use this:
```bash
sudo dnf install -y jq  python3-pip
sudo pip install yq
temp_dir=$(mktemp -d)
for fqdn in arthur.example.com ford.example.com trillian.example.com zaphod.example.com; do
  ssh ${fqdn} 'sudo certbot certificates' | sed -n '/Certificate Name:.*$/,/^[- ]+$/p' | head -n -1 | csplit --digits 2 --prefix ${temp_dir}/${fqdn}-cert- --suffix-format "%02d.txt" --elide-empty-files --quiet - "/Certificate Name:/0" "{*}"
  sed -i 's/  Certificate Name:/name:/g' ${temp_dir}/${fqdn}-cert-*.txt
  sed -i 's/    Serial Number:/serial:/g' ${temp_dir}/${fqdn}-cert-*.txt
  sed -i 's/    Key Type:/schema:/g' ${temp_dir}/${fqdn}-cert-*.txt
  sed -i 's/    Domains:/domains:/g' ${temp_dir}/${fqdn}-cert-*.txt
  sed -i 's/    Expiry Date:/expiry:/g' ${temp_dir}/${fqdn}-cert-*.txt
  sed -i 's/    Certificate Path:/cert:/g' ${temp_dir}/${fqdn}-cert-*.txt
  sed -i 's/    Private Key Path:/key:/g' ${temp_dir}/${fqdn}-cert-*.txt
  sed -i 's/ (VALID: [0-9]\+ days\?)//g' ${temp_dir}/${fqdn}-cert-*.txt
  yq --slurp '[.[] | .domains = (.domains | split(" "))]' ${temp_dir}/${fqdn}-cert-*.txt > ${temp_dir}/cert-list-${fqdn}.json
done
```

**Comment by rglover** [comment]
If you're working in JavaScript/Node.js give this a try. By no means bulletproof, but if you can get access to the output from `certbot certificates` as a string, this will get you most of the way there:

```javascript
const response = `Certificate Name: blah.blah.com
  Serial Number: xxx
  Key Type: RSA
  Domains: blah.blah.com
  Expiry Date: 2022-11-13 16:20:05+00:00 (VALID: 86 days)
  Certificate Path: /etc/letsencrypt/live/blah.blah.com/fullchain.pem
  Private Key Path: /etc/letsencrypt/live/blah.blah.com/privkey.pem`;

const json = response.split('\n')?.reduce((formatted, line) => {
  const trimmedLine = line?.trim();
  const [key, value] = trimmedLine?.split(': ');
  formatted[key] = value.trim().replace('(VALID', '').trim();
  return formatted;
}, {});

console.log(json);

// Generates...

{
    "Certificate Name": "blah.blah.com",
    "Serial Number": "xxx",
    "Key Type": "RSA",
    "Domains": "blah.blah.com",
    "Expiry Date": "2022-11-13 16:20:05+00:00",
    "Certificate Path": "/etc/letsencrypt/live/blah.blah.com/fullchain.pem",
    "Private Key Path": "/etc/letsencrypt/live/blah.blah.com/privkey.pem"
}
```

**Comment by kellyjonbrazil** [comment]
`jc` v1.23.2 now has a `certbot` parser which will convert `certbot show_account` and `certbot certificates` command output to JSON or YAML. UNIX Epoch and ISO timestamp fields are also added for convenience.

https://github.com/kellyjonbrazil/jc

```
% certbot show_account | jc --certbot -p
{
  "account": {
    "server": "https://acme-staging-v02.api.letsencrypt.org/directory",
    "url": "https://acme-staging-v02.api.letsencrypt.org/acme/acct/12345679",
    "email": "some@example.com"
  }
}
```
```
% certbot certificates | jc --certbot -p
{
  "certificates": [
    {
      "name": "example.com",
      "serial_number": "3f7axxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx",
      "key_type": "RSA",
      "domains": [
        "example.com",
        "www.example.com"
      ],
      "expiration_date": "2023-05-11 01:33:10+00:00",
      "validity": "63 days",
      "certificate_path": "/etc/letsencrypt/live/example.com/fullchain.pem",
      "private_key_path": "/etc/letsencrypt/live/example.com/privkey.pem",
      "expiration_date_epoch": 1683793990,
      "expiration_date_epoch_utc": 1683768790,
      "expiration_date_iso": "2023-05-11T01:33:10+00:00"
    },
    {
      "name": "example.org",
      "serial_number": "3bcyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyy",
      "key_type": "RSA",
      "domains": [
        "example.org",
        "www.example.org"
      ],
      "expiration_date": "2023-06-12 01:35:30+00:00",
      "validity": "63 days",
      "certificate_path": "/etc/letsencrypt/live/example.org/fullchain.pem",
      "private_key_path": "/etc/letsencrypt/live/example.org/privkey.pem",
      "expiration_date_epoch": 1686558930,
      "expiration_date_epoch_utc": 1686533730,
      "expiration_date_iso": "2023-06-12T01:35:30+00:00"
    }
  ]
}
```

**Comment by Ryuno-Ki** [comment]
I was today's years old when I learned about `jc` 😻 

As Debian Bookworm ships 1.22.5 I had to take a little detour.
Given the [announcement to sundown the email notifications](https://letsencrypt.org/2025/01/22/ending-expiration-emails/) I wrote a small shell script to list the expiration date for each domain I request a TLS cert for:

```sh
for cert in /etc/letsencrypt/live/*/cert.pem; do
    echo "$(ls $cert | cut -d '/' -f5): $(openssl x509 -in $cert -text | jc --x509-cert -p | jq '.[].tbs_certificate.validity.not_after_iso' | date +%F)";
done
```

Not as beautiful as your example but it gets the job done. I looked into how Certbot determines the expiration date and they also parse the `cert.pem` file using openssl (albeit programmatically).

Hope it helps! 🧡
